add stats api for internal metric collection

garrettbslone opened the pull request:
This adds a simple metric collection & sender api to `libflux` as suggested in issue #3517. The public api in `src/common/libflux/stats.h` defines the functions for storing metrics and changing the internal flush interval (which is `1s` by default). There are three types of supported metrics: `counters`, `gauges`, and `timers`. `Counters` send integers values which are aggregated in Brubeck over its flush interval. `Gauges` take an arbitrary integer value and maintain that value until next set, like a snapshot at a certain point. `Timers` take an amount of time in miliseconds. 

Right now there are no CI tests (something I need some suggestions on), but to test locally first set up Brubeck and Graphite which I have some instructions on setting up in my [https://github.com/flux-framework/fripp/blob/main/journals/week1.md](journal). Next, the stats will try getting the address to send to by reading the `FLUX_FRIPP_STATSD={IP}:{PORT}` environment variable. If you use the default config file from brubeck it should be set to `0.0.0.0:8126`. This branch has the stats running in the `content-cache`, but another use case is to use the [https://github.com/flux-framework/fripp/blob/main/code/jobtap-plugins/state-counts.c](state-counts plugin) which can be built like:

```bash
$ gcc -shared -I$HOME/flux-core -I$HOME/flux-core/src/include -fPIC \
    -o state-counts.so ../fripp.h ../fripp.c state-counts.c
```

and assuming that the plugin is in `../fripp/code/jobtap-plugins/`, it can be loaded into a running instance like this:

```bash
$ flux jobtap load $(pwd)/../fripp/code/jobtap-plugins/state-counts.so
```

garlick commented:
Maybe some variant of this would work as the basis for testing in sharness?
```
$  FLUX_FRIPP_STATSD=0.0.0.0:9999 src/cmd/flux start socat UDP-RECVFROM:9999 -
flux.content-cache.dirty:0|g
flux.content-cache.valid:10|g
flux.content-cache.size:1380|g
flux.content-cache.count:10|g
flux.content-cache.flush-batch-count:0|g
```

garrettbslone commented:
Force pushed to fix spelling of gauge in a few places..

grondo commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,67 @@
+ * the previous value. If 'inc' is set and 'name' was not previously stored, 
+ * then the value is stored directly.
+ */
+int flux_stats_gauge (flux_t *h, const char *name, ssize_t value, bool inc);
```
I think I argued for keeping the `inc` parameter here on the :coffee: call, but I was confused.

I do think there should be a way to increment, decrement, or set a gauge, but I would suggest/prefer 2 separate functions for clarity and ease of use in the code, e.g as a first cut, something like:

```c
int flux_stats_gauge_set (flux_t *h, const char *name, size_t val);
int flux_stats_gauge_inc (flux_t *h, const char *name, ssize_t val);
```

I guess `flux_stats_gauge_inc (h, "mystat", -1)` would then work to decrement a gauge by 1. Optionally, if preferred you could have a separate `flux_status_gauge_dec()`.

As we also talked about it might be nice to provide a "getter", but probably not necessary at this time.

(This means we could probably drop the `flux_stats_inc()` and `flux_stats_dec()` above?)

chu11 commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,438 @@
+/************************************************************\
+ * Copyright 2014 Lawrence Livermore National Security, LLC
```
just a reminder to update all copyright years to 2021 :-)

garlick commented on src/broker/content-cache.c:
```diff
@@ -953,6 +976,15 @@ struct content_cache *content_cache_create (flux_t *h, attr_t *attrs)
+              cache)))
+        goto error;
+    flux_watcher_start (cache->w);
+
```
Currently this timer fires unconditionally (whether or not we are collecting stats).  On a compute node, this might be stealing cycles from a job, or evicting cache lines, etc.. Is there a way we can disable the timer if we are not collecting stats?

I think it would be OK to piggy back on the `sync_cb()` which gets called every heartbeat (default T=2s).  See also: [flux_sync_create(3)](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man3/flux_sync_create.html).  Maybe just change the `timer_cb()` function into one that gets called from `sync_cb()`.

garlick commented on src/broker/content-cache.c:
```diff
@@ -963,4 +995,4 @@ struct content_cache *content_cache_create (flux_t *h, attr_t *attrs)
 /*
  * vi:tabstop=4 shiftwidth=4 expandtab
- */
+ */
```
Please fix trailing white space (I'll just call out this one - there are some others).

garlick commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,438 @@
+        flux_log_error (h, "fripp_ctx_create");
+        goto error;
+    }
+    if (!(uri = getenv ("FLUX_FRIPP_STATSD"))) {
+        flux_log_error (h, "FLUX_FRIPP_STATSD env var not set");
+        goto error;
+    }
```
Currently an error is logged if FLUX_FRIPP_STATSD is not set.  Moreover, it's logged for every metric, every time a stats function is called.  FRIPP should be silent if stats collection is not enabled.

Suggestion:  add a function for testing whether stats are enabled:
```c
bool fripp_enabled (flux_t *h, const char *metric);
```
This function can immediately return false if the environment variable is not set, else true.  Later we can use `metric` to test whether a particular metric prefix is enabled, but for now it could just be ignored and set to NULL by callers.

Then call it early to reduce overhead when stats collection is disabled.  For example, in the `flux_stats_*()` functions, return immediately, before trying to access/create context if `fripp_enabled(name)` return false.

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,55 @@
+
+/* Format a metric name and increment its value by '1'.
+ */
+int flux_stats_inc (flux_t *h, const char *fmt, ...)
```
Just wondering, should all these functions return `void`?  It seems unlikely that callers will want to be burdened with handling any potential error.

garrettbslone commented:
Force pushing to address some of the suggestions above. 

Starting with the `content-cache` what has changed is the `timer watcher` where the stats were being collected was removed, and they are now being collected inside of the `sync_cb`

In `stats.c/h`:
- removed the `flux_stats_inc/dec` functions
- removed the `inc` option from `flux_stats_gauge` and replaced it with two functions, one to set a gauge and one to increment one
- added a function to set an optional prefix for all metrics sent from the handle
- added a function to check if stats collection is enabled
- removed return value from stats functions
- pulled out some common code into a function
- if stats collection is not enabled return right away instead of trying to get `fripp_ctx`

In `fripp.c/h`:
- added a prefix to be prepended to all metrics in the `fripp_ctx` with a default of `flux.{rank}`
- added a function to check if stats collection is enabled

garlick commented on src/broker/content-cache.c:
```diff
@@ -920,6 +920,25 @@ void content_cache_destroy (struct content_cache *cache)
+                      flux_watcher_t *w, 
+                      int revents, 
+                      void *arg)
+{
```
Oops, should this be in this commit?

garlick commented on src/common/libflux/stats.c:
```diff
@@ -0,0 +1,109 @@
+        if (!(ctx = fripp_ctx_create (h)))
+            return NULL;
+        if (flux_aux_set (h, FRIPP_AUX_TAG, ctx, fripp_ctx_destroy) == -1)
+            return NULL;
```
leaks ctx on aux_set failure

garlick commented on src/common/libflux/stats.c:
```diff
@@ -0,0 +1,109 @@
+
+    char prefix[128];
+    if (vsnprintf (prefix, sizeof (prefix), fmt, ap) >= 128)
+        return;
```
be sure va_end() is called on all exit paths after va_start() is called.

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,79 @@
+ *           that are part of each flux_t handle. The counts will continually
+ *           increase and brubeck will handle calculating the count sent in
+ *           the interval.
+ * Gauge   - An integer that takes an arbitrary value and maintains its
+ *           it is set again. Gauges can also take incremental values
```
did you mean "maintains its value"?

Really appreciate the inline docs here!

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,79 @@
+void flux_stats_gauge_inc (flux_t *h, const char *name, ssize_t inc);
+
+
+/* Update (or create) and store 'ms' for 'name' to be sent on the
+ * next flush.
+ */
+void flux_stats_timing (flux_t *h, const char *name, double ts);
```
Comment references `ms` but argument is `ts`.  (Its not a timestamp, so ts might be misleading)

Name of function is actually `flux_stats_timer()` right?

garrettbslone replied:
Function is named `flux_stats_timing` (and `fripp_timing`) I believe to make it clear that fripp/stats don't keep track of the timer internally.

garlick replied:
My bad, I thought I read "timer" in the C code and "timing" in the header.  You're right.

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,79 @@
+ */
+void flux_stats_timing (flux_t *h, const char *name, double ts);
+
+/* Update the internal aggregation period over which metrics accumulate
+ * before being set. A 'period' of '0' indicates the metrics should be
+ * sent immediately.
+ */
```
might be useful to state the default aggregation period here.

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,79 @@
+ */
+void flux_stats_set_period (flux_t *h, double period);
+
+/* Set the prefix {prefix}.{metric} to be preprended to all metrics
+ * sent from the handle. The prefix has a max limit of 127 characters.
+ * The default prefix is flux.{rank}.
+ */
+void flux_stats_set_prefix (flux_t *h, const char *fmt, ...);
```
We use mustache templates elsewhere in Flux so the curly braces might be a bit confusing.  For a second there I thought maybe `{rank}` would be expanded if I included it in the prefix.  Hmm, should it?

Also "set the {prefix}.{metric}" is a bit confusing.  Aren't we just setting the prefix?

grondo replied:
>  I thought maybe {rank} would be expanded if I included it in the prefix. Hmm, should it?

Mustache templates have double curly braces, e.g. `{{rank}}`, but I agree that curly braces could be confusing (esp. when you've recently worked in Python :wink:).

I like the idea of allowing expansion in metric names to ease the burden of adding metrics to any site. W
We do have a bare bones mustache renderer currently in the flux-shell source [here](https://github.com/flux-framework/flux-core/blob/master/src/shell/mustache.h). We could eventually move that into libutil for use in other Flux code.

However, it seems fine to leave template expansion out of this particular PR IMO.

garlick replied:
Oh yeah :-)  I was thinking of `flux jobs --format`.  Agreed, expansion could be done in another PR, if we want to go that way.

garlick commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,79 @@
+ */
+void flux_stats_set_prefix (flux_t *h, const char *fmt, ...);
+
+bool flux_stats_enabled (flux_t *h, const char *metric);
```
Missing header doc?

garlick commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,456 @@
+    double period;
+};
+
+bool fripp_enabled (flux_t *h, const char *metric)
```
Why does this function take a `flux_t` handle?

garrettbslone replied:
It takes a `flux_t` handle instead of a `fripp_ctx` to avoid trying to get/create a `fripp_ctx` each time it checks if collection is on.

garrettbslone replied:
Maybe instead it should take a`fripp_ctx` and then if the `fripp_ctx` is `NULL` it can check if it's enabled on the instance.

garlick replied:
That function isn't using the `flux_t` though:
```c
bool fripp_enabled (flux_t *h, const char *metric)
{
    return getenv ("FLUX_FRIPP_STATSD") != NULL;
}
```

garlick commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,456 @@
+#include "src/common/libczmqcontainers/czmq_containers.h"
+#include "src/common/libutil/iterators.h"
+
+#define FRIPP_MAX_PACKET_LEN 1440
+#define INTERNAL_BUFFSIZE 128
```
Shouldn't the buffer be larger than the max packet size?

garrettbslone replied:
The buffer is dynamically sized. It is initialized to the `INTERNAL_BUFFSIZE`, but that size is also used for the formatting of metrics in `fripp_packet_appendf` and for the `prefix` buffer.

garlick replied:
Ah OK, sorry I was confused there.

garrettbslone replied:
Maybe it could benefit from a better name or a comment?

garlick commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,456 @@
+            return -1;
+        if (sendto (ctx->sock, packet, len, 0,
+                    (void *) &ctx->si_server, sock_len) < 0)
+            fprintf (stderr, "packet %s dropped: %s\n", packet,
+                     strerror (errno));
```
the fprintf to stderr should be dropped (leftover debugging?)

It's not great practice (though sometimes the only way) to print stuff on stdio from a library.  Here the entire stats collection mechanism ought to just be "best effort" IMHO.

garrettbslone replied:
Should the `fprintf` on line 314 (in the `timer_cb`) also be dropped?

garlick replied:
Missed that one.  Yep, IMHO.

garlick reviewed (commented):
A few comments based on a quick skim.  I'm headed out on a walk and will be back later, but wanted to get these out there.

grondo commented on t/t1200-stats-basic.t:
```diff
@@ -0,0 +1,14 @@
+test_under_flux 2
+
+test_expect_success 'content-cache packets recieved' '
+	FLUX_FRIPP_STATSD=0.0.0.0:9999 flux start socat UDP-RECVFROM:9999 -
```
Oops, `socat` is not installed in the ci images (nor on my system), so this test is failing.

Also, how will the `socat` know to exit so the test stops?

Here's one idea. You can create a UDP listener in Python for use with the tests. Here's a basic demo that waits for just one packet and exits. (You'd want to add a timeout to this test to detect no packets and fail.

This version doesn't check the format of the packet, but this could easily be added. The script or different scripts could be enhanced to check for properly formatted or expected stats strings, so it might be a good base to start from:

```diff
diff --git a/t/t1200-stats-basic.t b/t/t1200-stats-basic.t
index 1e9f2aa3f..2bfef452b 100755
--- a/t/t1200-stats-basic.t
+++ b/t/t1200-stats-basic.t
@@ -7,8 +7,18 @@ test_description='Test stats collection and sending'
 
 test_under_flux 2
 
+test_expect_success 'create udp client program' '
+       cat <<-EOF >listen.py &&
+       import socket
+       s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+       s.bind(("0.0.0.0", 9999))
+       print(s.recvfrom(1024))
+       EOF
+       chmod +x listen.py
+'
+
 test_expect_success 'content-cache packets recieved' '
-       FLUX_FRIPP_STATSD=0.0.0.0:9999 flux start socat UDP-RECVFROM:9999 -
+       FLUX_FRIPP_STATSD=0.0.0.0:9999 flux start flux python listen.py
 '
 
 test_done

```

grondo replied:
```
 grondo@asp:~/git/f.git/t$ ./t1200-stats-basic.t  -d -v
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/01-setup.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/flux-sharness.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/01-setup.sh
sharness: loading extensions from /home/grondo/git/f.git/t/sharness.d/flux-sharness.sh
expecting success: 
	cat <<-EOF >listen.py &&
	import socket
	s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
	s.bind(("0.0.0.0", 9999))
	print(s.recvfrom(1024))
	EOF
	chmod +x listen.py

ok 1 - create udp client program

expecting success: 
	FLUX_FRIPP_STATSD=0.0.0.0:9999 flux start flux python listen.py

(b'flux.0.content-cache.size:1360|g\nflux.0.content-cache.dirty:0|g\nflux.0.content-cache.count:10|g\nflux.0.content-cache.flush-batch-count:0|g\nflux.0.content-cache.valid:10|g', ('127.0.0.1', 46371))
ok 2 - content-cache packets recieved

# passed all 2 test(s)
1..2

```

garrettbslone replied:
`socat` was exiting after receiving the first packet when I was testing. Not sure if it's safe to assume that'll always happen.

grondo replied:
Ah, apologies I didn't know that (and didn't have socat to test with :wink:)

If we can get the python script to parse and validate the packets, then that will really solidify the testing.
You could add options to wait for N packets, or wait for a value to appear (e.g. if you wanted to ensure 1 running job was counted, etc.)

grondo commented on src/common/libflux/stats.h:
```diff
@@ -0,0 +1,82 @@
+ *           An example of where to use a gauge is to track the current
+ *           size of the broker's content-cache. At each point, the cache's
+ *           sizes are independent of each other.
+ * Timer   - A double value which represents the time taken for a given
```
I think @garlick had already commented on this, but I can't find it.

The function to set a "timer" metric is called `flux_stats_timing()`, but the docs here reference "Timer". Should one of the two be updated? Maybe just document this stat as a "Timing" metric?

grondo commented on t/t1200-stats-basic.t:
```diff
@@ -0,0 +1,92 @@
+export FLUX_FRIPP_STATSD=0.0.0.0:9999
+
+test_expect_success 'create udp client program' '
+	cat <<-EOF >listen.py &&
```
Since this script is larger than a few lines now, feel free to include it directly in `t/scripts/stats-listen.py` or something to ease future development. (Sorry, this is my fault, I should have suggested that initially)

grondo commented on t/t1200-stats-basic.t:
```diff
@@ -0,0 +1,92 @@
+
+	s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+	s.bind(("0.0.0.0", 9999))
+	s.settimeout(args.timeout)
```
Instead of trying to internally track the timeout within `listen.py` it may be easier to run the program under our builtin `run_timeout` function. Then you don't have to worry about  tracking an overall timeout with blocking network code, which is difficult.

You could still keep the `-w, --wait-for=N` and `-s, --search-for=METRIC` options, just exit with zero status after N packets are received or you find the matching metric. On timeout, `run_timeout` will kill the script and trigger a failure.

grondo commented:
I just threw in a couple on-off comments on the `listen.py` test program, which may be helpful to get it working in CI.

Also, I noticed a typo in the commit message for b6991a00fe5918128ca8f7e02001584ab73c0813 ("validtion")

grondo commented:
It case it isn't clear why the `python format` check is failing for you -- we use the `black` formatter to consistently format all Python, even test scripts in `t/scripts`. Install the required version of `black` with `python3 -m pip install 'black==20.08.b1' --force-reinstall`, and then either run `black` on your script or run `./scripts/format` to run the formatter on all python scripts.

garrettbslone commented:
squashed down the commit history and force pushed

grondo commented on t/stats/stats-immediate.c:
```diff
@@ -0,0 +1,79 @@
+
+    flux_stats_set_prefix (h, "flux.job.state.immediate");
+    flux_stats_set_period (h, 0.0);
+    
```
Whitespace error here and at end of file.

grondo commented on t/stats/stats-immediate.c:
```diff
@@ -0,0 +1,79 @@
+    flux_stats_set_prefix (h, "flux.job.state.immediate");
+    flux_stats_set_period (h, 0.0);
+    
+    flux_stats_set_prefix (h, "aQmi173rvgumStDdMZdwtJtpLLVJOUXol2aDndev/XsH/gM\
+wlPz/vMZhajJWGctwJZa1uFoAoINjwITPvezGoQDb/9DD3vkPcknN+f/u3vSc0tg/+3aFTONhUIomK\
+B4qiSKSotbtZl3Ewe2Oh+wI+nuG3/ebqIXSoEXjIFOB7vvGA==");
```
What is the purpose of this interesting prefix? If nothing else, I'd suggest at least a comment to help future code spelunkers.

grondo replied:
Also I notice prefix is already set above in L71, maybe this was debugging code?

garrettbslone replied:
This was to make sure that a prefix longer than the limit (127 characters) wouldn't be set and leave the prefix unchanged.

grondo replied:
Ah! Thanks :-)

grondo commented on t/stats/stats-immediate.c:
```diff
@@ -0,0 +1,79 @@
+    flux_t *h;
+    if (!(h = flux_jobtap_get_flux (p)))
+        return -1;
+    if (!(data = calloc (1, sizeof (*data))))
```
`data` is leaked. You can tie its destruction to the plugin with `flux_plugin_aux_set(3)`

grondo commented on src/common/libflux/fripp.c:
```diff
@@ -0,0 +1,468 @@
+#define DEFAULT_AGG_PERIOD 1.0
+
+typedef enum {
+    BRUBECK_COUNTER,
+    BRUBECK_GAUGE,
+    BRUBECK_TIMER
```
Question: would these be better named STATSD, or are these metric types specific to brubeck?

garrettbslone replied:
Well, they are all Statsd types and the gauge and timer use the same syntax for brubeck and statsd, but the counter uses a slightly different syntax for brubeck (`C`) than statsd (`c`).

garrettbslone replied:
maybe better for consistency for now?

grondo replied:
This is all internal to `fripp.c` so I guess I don't have a strong feeling at this point.

grondo commented on t/stats/stats-basic.c:
```diff
@@ -0,0 +1,73 @@
+    flux_t *h;
+    if (!(h = flux_jobtap_get_flux (p)))
+        return -1;
+    if (!(data = calloc (1, sizeof (*data))))
```
`data` is leaked (tie to plugin with `flux_plugin_aux_set()`)

grondo reviewed (commented):
Ok, just a few comments this round.

Also noticed a typo in the PR title: "interal" missing `n`

garrettbslone commented:
Force pushing to address comments above, removed wip tag from title, and fixed spelling in title.

grondo commented on t/t1200-stats-basic.t:
```diff
@@ -0,0 +1,54 @@
+plugin_i=${FLUX_BUILD_DIR}/t/stats/.libs/stats-immediate.so
+plugin_b=${FLUX_BUILD_DIR}/t/stats/.libs/stats-basic.so
+
+test_expect_success 'load jobtap plugin' '
+	flux jobtap load $plugin_i && flux jobtap load $plugin_b &&
+	flux jobtap list > plugins &&
+	grep stats-basic plugins && grep stats-immediate plugins
+'
```
One more comment. I don't think we need this test to run under `test_under_flux`, nor have the jobtap plugins loaded here. I may have missed something but the rest of the functionality tests all use `flux start`, so the "outer" instance created by `test_under_flux` actually isn't used, nor are the loaded plugins.

If that is true then you coudl just remove `test_under_flux 2` and the `load jobtap plugin` test.

grondo commented:
Might as well also rebase against current master. :grin:

grondo commented:
BTW, you could probably quiet that LGTM error by binding to `127.0.0.1` instead of `0.0.0.0`.

grondo reviewed (approved):
LGTM! Nice work @garrettbslone!

garlick commented:
Yes thanks for all your hard work this summer.  The finished produce is great!

Ready to merge?  If so I'll set MWP and babysit that tempormental coverage test.

garrettbslone commented:
I think it's ready to merge!

garlick commented:
Oops I had just set MWP before seeing your comments @grondo.  OK, @garrett, want to take care of that change first?

garlick commented:
Setting it again - sorry, I was looking at an older comment.

garlick commented:
@mergifyio rebase

grondo commented:
:tada: nice work @garrettbslone!
